src/commands/redis_cmd.cc

```diff
@@ -2554,152 +2556,135 @@ class CommandZPopMax : public CommandZPop {
 class CommandZRange : public Commander {
  public:
   explicit CommandZRange(bool reversed = false) : reversed_(reversed) {}
-
   Status Parse(const std::vector<std::string> &args) override {
-    auto parse_start = ParseInt<int>(args[2], 10);
-    auto parse_stop = ParseInt<int>(args[3], 10);
-    if (!parse_start || !parse_stop) {
-      return Status(Status::RedisParseErr, errValueNotInteger);
+    int offset = 0;
+    int count = -1;
+    CommandParser parser(args, 4);
+    while (parser.Good()) {
+      if (parser.EatEqICaseFlag("BYSCORE", by_flag_)) {
+        spec_ = std::make_shared<ZRangeSpec>();
```

Comment:
Could you explain a bit about the use of `shared_ptr` here?

Reply:
Concurrent instructions result in destructions on line `auto s = svr_->LookupAndCreateCommand(cmd_tokens.front(), &current_cmd_);`.
```
void Connection::ExecuteCommands(std::deque<CommandTokens> *to_process_cmds) {
  Config *config = svr_->GetConfig();
  std::string reply, password = config->requirepass;

  while (!to_process_cmds->empty()) {
    auto cmd_tokens = to_process_cmds->front();
    to_process_cmds->pop_front();

    if (IsFlagEnabled(Redis::Connection::kCloseAfterReply) && !IsFlagEnabled(Connection::kMultiExec)) break;

    auto s = svr_->LookupAndCreateCommand(cmd_tokens.front(), &current_cmd_);
```

Reply:
I still can't understand this point, the lookup command is unique ptr and it should not have the race between workers.

Reply:
> I still can't understand this point, the lookup command is unique ptr and it should not have the race between workers.

There are two tasks `t1`and `t2` at the same time. `t2` uses the same `current_cmd_` to get allocation space while t1 is still executing, which causes t1 to use the resources held by `unique_ptr` to be freed.

Reply:
It is not correct. If `CommandZrange` has this problem, then other commands will also have this problem.

Reply:
I think union may have similar problem: we need to determine the correct dtor since there are some non-POD in spec_ (i.e. std::string in LexSpec), maybe we can just use multiple value or std::variant here.